### DISCUSSION TITLE
Translator

### DISCUSSION THREAD

**Discussion by mi66mc** [Rapptz/discord.py#9160] [Q&A] [answered]
Can someone show me an translator example using `app_commands.Translator`? It will help me a lot!

**Comment by cibere** [comment]
~~If you head over to the [dpy discord server](https://discord.com/invite/dpy), `?tag localization` has an example.~~

**Comment by Rapptz** [comment, marked as answer]
The general idea for translation is hopefully simple, first you need to make an `app_commands.Translator` subclass that contains the necessary implementation for dealing with your localisation. For example:

```py
from __future__ import annotations

import discord
from discord import app_commands

class MyTranslator(app_commands.Translator):
    async def translate(
        self,
        string: app_commands.locale_str,
        locale: discord.Locale,
        context: app_commands.TranslationContext,
    ) -> str | None:
        # For this example, we can translate a few words in Japanese...
        message = str(string)
        if locale is discord.Locale.japanese:
            if message == 'Hello!':
                return 'こんにちは！'
            elif message == 'Goodbye!':
                return 'さようなら！'

        # Otherwise we don't handle it
        return None
```

Obviously a more complicated implementation of `Translator` would work with things such as gettext or fluent, but for now this simple instance is enough to get it started.

Afterwards you set the translator using `tree.set_translator`, e.g. `await tree.set_translator(MyTranslator())` then afterwards you call `await tree.sync()` as usual for the translations to show up on Discord.

**Reply by mi66mc** [reply]
Thank you so much!
